packages/lib-classifier/src/store/SubjectStore.spec.js

```diff
@@ -1,147 +1,131 @@
-          subjects
-        }
-      })
-    }
-  }
-}
-
-describe('Model > SubjectStore', function () {
-  function setupStores (rootStore) {
-    sinon.stub(rootStore.classifications, 'createClassification')
-    rootStore.projects.setResource(project)
-    rootStore.workflows.setResource(workflow)
-    rootStore.workflows.setActive(workflow.id)
+    store.projects.setResource(project)
+    store.projects.setActive(project.id)
+    store.workflows.setResource(workflow)
+    store.workflows.setActive(workflow.id)
+    return store
   }
   describe('Actions > advance', function () {
+    let rootStore
     before(function () {
-      rootStore = RootStore.create(
-        { projects: ProjectStore.create(), subjects: SubjectStore.create(), workflows: WorkflowStore.create() },
-        { client: clientStub }
-      )
+      rootStore = setupStores()
     })
 
-    it('should make the next subject in the queue active when calling `advance()`', function (done) {
-      setupStores(rootStore)
-      rootStore.subjects.populateQueue().then(() => {
-        expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
-        rootStore.subjects.advance()
-        expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
-        expect(rootStore.subjects.resources.get('1')).to.be.undefined()
-      }).then(done, done)
+    it('should make the next subject in the queue active when calling `advance()`', function () {
+      expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
+      rootStore.subjects.advance()
+      expect(rootStore.subjects.active.id).to.equal(rootStore.subjects.resources.values().next().value.id)
+      expect(rootStore.subjects.resources.get('1')).to.be.undefined()
     })
 
     describe('with less than three subjects in the queue', function () {
-      let populateSpy
+      let rootStore
 
       before(function () {
-        populateSpy = sinon.spy(rootStore.subjects, 'populateQueue')
+        rootStore = setupStores()
+      })
+
+      beforeEach(function () {
+        rootStore.subjects.populateQueue.resetHistory()
         while (rootStore.subjects.resources.size > 2) {
           rootStore.subjects.advance()
         }
       })
 
+      afterEach(function () {
+      })
+
       after(function () {
         rootStore.subjects.populateQueue.restore()
       })
 
       it('should request more subjects', function () {
-        expect(populateSpy).to.have.been.calledOnce()
+        expect(rootStore.subjects.populateQueue).to.have.been.calledOnce()
+      })
+
+      it('should request more subjects when the initial queue response only has less than 3 subjects', function () {
+        const clientStub = stubPanoptesJs({ workflows: workflow, subjects: shortListSubjects })
+        rootStore = setupStores(clientStub)
+        expect(rootStore.subjects.populateQueue).to.have.been.calledOnce()
```

Comment:
I think `populateQueue` should be called twice here: once when the workflow is attached (to prepopulate the queue), then again when the first active subject is advanced (because the initial queue is only 2 subjects.)

Reply:
Should this test call `rootStore.subjects.advance()` or does that happen automatically after the workflow is attached?

Reply:
I'm resetting the history before the expectation in the `beforeEach` because this is only testing for the behavior to call `populateQueue` again. I thought it'd be confusing to test for both calls, since that's not being stated here as being tested.

Reply:
Is that because this block just observes what happens when we drop an existing queue from 3 to 2 subjects?

What might be missing here is that the `beforeEach` block doesn't repopulate the queue before advancing it down to 2 subjects, so that behaviour is only tested in the first test. It's hard to tell, though. This new test sets up a new, empty queue which will then populate when the workflow attaches and you might be testing for the initial call that happens then.

Reply:
I've double-checked, and the queue is empty at this point in the test, so the expectation is being checked before the API promise has resolved with 2 new subjects.

Reply:
I forgot about the asynchronicity of the initial store setup myself here, so moving this into its own describe block with before function setup should not only improve readability but ensure that we're testing what we think here.